File: src/Fable.Repl.Lib/Feliz.Snabbdom/Snabbdom.fs
module Snabbdom

open System
open Fable.Core

type Module = interface end

type VNode =
    abstract sel: string with get, set
    abstract key: string with get, set
    abstract data: obj with get, set
    abstract children: VNode[] with get, set
    abstract text: string with get, set
    abstract elm: Browser.Types.HTMLElement with get, set

type Patch = delegate of VNode * VNode -> VNode

// [<ImportMember("snabbdom/modules/attributes")>]
[<ImportMember("${outPath}/snabbdom.min.js")>]
let attributesModule: Module = jsNative

// [<ImportMember("snabbdom/modules/style")>]
[<ImportMember("${outPath}/snabbdom.min.js")>]
let styleModule: Module = jsNative

// [<ImportMember("snabbdom/modules/eventlisteners")>]
[<ImportMember("${outPath}/snabbdom.min.js")>]
let eventListenersModule: Module = jsNative

// [<ImportMember("snabbdom/h")>]
[<ImportMember("${outPath}/snabbdom.min.js")>]
let h(tag: string, props: obj, children: ResizeArray<VNode>): VNode = jsNative

// [<ImportMember("snabbdom/init")>]
[<ImportMember("${outPath}/snabbdom.min.js")>]
let init(modules: Module[]): Patch = jsNative

[<ImportMember("${outPath}/snabbdom.min.js")>]
let memo(key: Guid, render: 'arg -> VNode, arg: 'arg, equals: ('arg -> 'arg -> bool) option): VNode = jsNative

type Helper() =
    static let patcher = init [|
        attributesModule
        styleModule
        eventListenersModule
    |]
    static member Empty: VNode = unbox null
    static member Text(str: string): VNode = unbox str
    static member Patch(oldNode: VNode, newNode: VNode): VNode = patcher.Invoke(oldNode, newNode)
    static member Patch(el: Browser.Types.HTMLElement, vnode: VNode): VNode = patcher.Invoke(unbox el, vnode)
    static member Memo(key: Guid, render: 'Arg -> VNode, arg: 'Arg, ?equals: 'Arg -> 'Arg -> bool): VNode = memo(key, render, arg, equals)

File: js/repl/fable-repl-lib/Feliz.Snabbdom/Snabbdom.js
import { class_type } from "../../fable-library/Reflection.js";
import { memo, eventListenersModule, styleModule, attributesModule, init } from "./snabbdom.min.js";

export class Snabbdom_Patch {
    constructor() {
    }
}

export function Snabbdom_Patch$reflection() {
    return class_type("Snabbdom.Patch", void 0, Snabbdom_Patch, class_type("System.MulticastDelegate"));
}

export class Snabbdom_Helper {
    constructor() {
    }
}

export function Snabbdom_Helper$reflection() {
    return class_type("Snabbdom.Helper", void 0, Snabbdom_Helper);
}

export function Snabbdom_Helper_$ctor() {
    return new Snabbdom_Helper();
}

(() => {
    Snabbdom_Helper.patcher = init([attributesModule, styleModule, eventListenersModule]);
})();

export function Snabbdom_Helper_get_Empty() {
    return null;
}

export function Snabbdom_Helper_Text_Z721C83C5(str) {
    return str;
}

export function Snabbdom_Helper_Patch_7B47CB60(oldNode, newNode) {
    return Snabbdom_Helper.patcher(oldNode, newNode);
}

export function Snabbdom_Helper_Patch_Z7AA7ED5A(el, vnode) {
    return Snabbdom_Helper.patcher(el, vnode);
}

export function Snabbdom_Helper_Memo_Z25D8C161(key, render, arg, equals) {
    return memo(key, render, arg, equals);
}

